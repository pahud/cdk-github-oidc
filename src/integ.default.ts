import {
  Stack, App,
} from 'aws-cdk-lib';
import { Provider } from './';

const app = new App();

const env = {
  region: process.env.CDK_DEFAULT_REGION,
  account: process.env.CDK_DEFAULT_ACCOUNT,
};

const stack = new Stack(app, 'demo-stack', { env });

const provider = new Provider(stack, 'GithubOpenIdConnectProvider');
provider.createRole('gh-oidc-role',
  [
    { owner: 'pahud', repo: 'gitpod-workspace' },
    { owner: 'pahud', repo: 'github-codespace' },
    { owner: 'pahud', repo: 'vscode' },
  ],
);