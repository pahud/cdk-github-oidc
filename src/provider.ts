import {
  Resource, Stack,
  aws_iam as iam,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

const DEFAULTS: { [key: string]: string } = {
  issuer: 'token.actions.githubusercontent.com',
  thumbprint: 'a031c46782e6e6c662c2c87c76da9aa62ccabd8e',
};

/**
 * Represents a GitHub repository
 */
export interface RepositoryConfig {
  /**
   * The owner of the repository.
   */
  readonly owner: string;
  /**
   * The name of the repository.
   */
  readonly repo: string;
  /**
   * The sub prefix string from the JWT token used to be validated by AWS. Appended after `repo:${owner}/${repo}:`
   * in an IAM role trust relationship. The default value '*' indicates all branches and all tags from this repo.
   *
   * Example:
   * repo:octo-org/octo-repo:ref:refs/heads/demo-branch - only allowed from `demo-branch`
   * repo:octo-org/octo-repo:ref:refs/tags/demo-tag - only allowed from `demo-tag`.
   * repo:octo-org/octo-repo:pull_request - only allowed from the `pull_request` event.
   * repo:octo-org/octo-repo:environment:Production - only allowd from `Production` environment name.
   *
   * @default '*'
   * @see https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#configuring-the-oidc-trust-with-the-cloud
   */
  readonly filter?: string;
}

export abstract class ProviderBase extends Resource {
  public abstract readonly openIdConnectProvider: iam.IOpenIdConnectProvider;
  /**
   *
   * @param repo a list of repositories
   * @returns a list of subjects
   */
  private formatSubject(repo: RepositoryConfig[]): string[] {
    return repo.map(r => `repo:${r.owner}/${r.repo}:${r.filter ?? '*'}`);
  }
  public createRole(id: string, repo: RepositoryConfig[], roleProps?: iam.RoleProps):iam.Role {
    if (repo.length == 0) {
      throw new Error('Error - at least one repository is required');
    }
    const role = new iam.Role(this, id, {
      ...roleProps,
      assumedBy: new iam.OpenIdConnectPrincipal(this.openIdConnectProvider, {
        StringLike: {
          [`${DEFAULTS.issuer}:sub`]: this.formatSubject(repo),
        },
      }),
    });
    return role;
  }
}

/**
 * The Github OpenID Connect Provider
 */
export class Provider extends ProviderBase {
  /**
   * import the existing provider
   */
  public static fromAccount(scope: Construct, id: string): ProviderBase {
    class Import extends ProviderBase {
      public readonly openIdConnectProvider: iam.IOpenIdConnectProvider;
      constructor(s: Construct, i: string) {
        super(s, i);
        // arn:aws:iam::xxxxxxxxxxxx:oidc-provider/token.actions.githubusercontent.com
        const arn = Stack.of(scope).formatArn({
          resource: 'oidc-provider',
          service: 'iam',
          region: '',
          resourceName: DEFAULTS.issuer,
        });
        this.openIdConnectProvider = iam.OpenIdConnectProvider.fromOpenIdConnectProviderArn(scope, `Provider${id}`, arn);
      }
    }
    return new Import(scope, id);
  }
  public readonly issuer: string = DEFAULTS.issuer;
  public readonly openIdConnectProvider: iam.IOpenIdConnectProvider;
  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.openIdConnectProvider = new iam.OpenIdConnectProvider(this, `Provider${id}`, {
      url: `https://${DEFAULTS.issuer}`,
      clientIds: ['sts.amazonaws.com'],
      thumbprints: [`${DEFAULTS.thumbprint}`],
    });
  }
}