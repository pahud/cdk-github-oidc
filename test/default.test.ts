import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as cdk from '@aws-cdk/core';
import { Provider } from '../src';

let app: cdk.App;
let stack: cdk.Stack;

beforeEach(() => {
  app = new cdk.App();
  stack = new cdk.Stack(app, 'demo-stack');
});

test('create a default provider', () => {
  // GIVEN
  // WHEN
  new Provider(stack, 'Provider');
  // match snapshot
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  // we should have the provider
  expect(stack).toHaveResource('Custom::AWSCDKOpenIdConnectProvider', {
    ServiceToken: {
      'Fn::GetAtt': [
        'CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0',
        'Arn',
      ],
    },
    ClientIDList: [
      'sts.amazonaws.com',
    ],
    ThumbprintList: [
      'a031c46782e6e6c662c2c87c76da9aa62ccabd8e',
    ],
    Url: 'https://token.actions.githubusercontent.com',
  });
});

test('create iam role for single repository', () => {
  // GIVEN
  const provider = new Provider(stack, 'Provider');
  // WHEN
  provider.createRole('gh-oidc-role',
    [
      { owner: 'octo-org', repo: 'repo' },
    ],
  );
  // we should have a correct IAM role
  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRoleWithWebIdentity',
          Condition: {
            StringLike: {
              'token.actions.githubusercontent.com:sub': [
                'repo:octo-org/repo:*',
              ],
            },
          },
          Effect: 'Allow',
          Principal: {
            Federated: {
              Ref: 'ProviderProviderProviderBCE1C93C',
            },
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
});

test('create iam role for single repository specific branch', () => {
  // GIVEN
  const provider = new Provider(stack, 'Provider');
  // WHEN
  provider.createRole('gh-oidc-role',
    [
      { owner: 'octo-org', repo: 'repo', filter: 'ref:refs/heads/demo-branch' },
    ],
  );
  // we should have a correct IAM role and `StringLike` condition.
  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRoleWithWebIdentity',
          Condition: {
            StringLike: {
              'token.actions.githubusercontent.com:sub': [
                'repo:octo-org/repo:ref:refs/heads/demo-branch',
              ],
            },
          },
          Effect: 'Allow',
          Principal: {
            Federated: {
              Ref: 'ProviderProviderProviderBCE1C93C',
            },
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
});

test('create iam role for single repository specific tag', () => {
  // GIVEN
  const provider = new Provider(stack, 'Provider');
  // WHEN
  provider.createRole('gh-oidc-role',
    [
      { owner: 'octo-org', repo: 'repo', filter: 'ref:refs/tags/demo-tag' },
    ],
  );
  // we should have a correct IAM role and `StringLike` condition.
  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRoleWithWebIdentity',
          Condition: {
            StringLike: {
              'token.actions.githubusercontent.com:sub': [
                'repo:octo-org/repo:ref:refs/tags/demo-tag',
              ],
            },
          },
          Effect: 'Allow',
          Principal: {
            Federated: {
              Ref: 'ProviderProviderProviderBCE1C93C',
            },
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
});

test('create iam role for multiple repositories', () => {
  // GIVEN
  const provider = new Provider(stack, 'Provider');
  // WHEN
  provider.createRole('gh-oidc-role',
    [
      { owner: 'octo-org', repo: 'first-repo' },
      { owner: 'octo-org', repo: 'second-repo' },
      { owner: 'octo-org', repo: 'third-repo' },
    ],
  );
  // we should have a correct IAM role
  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRoleWithWebIdentity',
          Condition: {
            StringLike: {
              'token.actions.githubusercontent.com:sub': [
                'repo:octo-org/first-repo:*',
                'repo:octo-org/second-repo:*',
                'repo:octo-org/third-repo:*',
              ],
            },
          },
          Effect: 'Allow',
          Principal: {
            Federated: {
              Ref: 'ProviderProviderProviderBCE1C93C',
            },
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
});

test('create an iam role from the imported provider', () => {
  // GIVEN
  const provider = Provider.fromAccount(stack, 'Provider');
  // WHEN
  provider.createRole('gh-oidc-role',
    [
      { owner: 'octo-org', repo: 'repo' },
    ],
  );
  // we should have a correct IAM role
  expect(stack).toHaveResource('AWS::IAM::Role', {
    AssumeRolePolicyDocument: {
      Statement: [
        {
          Action: 'sts:AssumeRoleWithWebIdentity',
          Condition: {
            StringLike: {
              'token.actions.githubusercontent.com:sub': [
                'repo:octo-org/repo:*',
              ],
            },
          },
          Effect: 'Allow',
          Principal: {
            Federated: {
              'Fn::Join': [
                '',
                [
                  'arn:',
                  {
                    Ref: 'AWS::Partition',
                  },
                  ':iam::',
                  {
                    Ref: 'AWS::AccountId',
                  },
                  ':oidc-provider/token.actions.githubusercontent.com',
                ],
              ],
            },
          },
        },
      ],
      Version: '2012-10-17',
    },
  });
});

test('throw with empty repo', () => {
  // GIVEN
  const provider = new Provider(stack, 'Provider');
  // WHEN
  // we should throw an error
  expect(() => {
    provider.createRole('gh-oidc-role',
      [
      // empty repo
      ],
    );
  }).toThrow(/Error - at least one repository is required/);
});