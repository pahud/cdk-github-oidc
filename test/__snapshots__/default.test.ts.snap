// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a default provider 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850ArtifactHash23C38986": Object {
      "Description": "Artifact hash for asset \\"691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850\\"",
      "Type": "String",
    },
    "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850S3BucketDBD704BF": Object {
      "Description": "S3 bucket for asset \\"691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850\\"",
      "Type": "String",
    },
    "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850S3VersionKey7CB5D9CD": Object {
      "Description": "S3 key for asset version \\"691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850S3BucketDBD704BF",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850S3VersionKey7CB5D9CD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters691fcf2f421cd698a78dd390261a7120c50bef674fec27c16e9772e758d8f850S3VersionKey7CB5D9CD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ProviderProviderProviderBCE1C93C": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "sts.amazonaws.com",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "ThumbprintList": Array [
          "a031c46782e6e6c662c2c87c76da9aa62ccabd8e",
        ],
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
